File: core/src/com/unciv/logic/battle/air_interception.py
import random
from typing import Optional, Sequence, Tuple
from dataclasses import dataclass

@dataclass
class DamageDealt:
    attacker_dealt: int
    defender_dealt: int

    @classmethod
    def none(cls) -> 'DamageDealt':
        return cls(0, 0)

class MapUnitCombatant:
    def __init__(self, unit):
        self.unit = unit

    def get_name(self) -> str:
        return self.unit.name

    def get_civ_info(self):
        return self.unit.civ

    def get_tile(self):
        return self.unit.current_tile

    def is_defeated(self) -> bool:
        return self.unit.health <= 0

    def take_damage(self, damage: int):
        self.unit.health = max(0, self.unit.health - damage)

class AirInterception:
    @staticmethod
    def air_sweep(attacker: MapUnitCombatant, attacked_tile) -> None:
        # Air Sweep counts as an attack, even if nothing else happens
        attacker.unit.attacks_this_turn += 1

        # Use up movement
        if (attacker.unit.has_unique("CanMoveAfterAttacking") or
            attacker.unit.max_attacks_per_turn() > attacker.unit.attacks_this_turn):
            if not attacker.unit.base_unit.moves_like_air_units:
                attacker.unit.use_movement_points(1.0)
        else:
            attacker.unit.current_movement = 0.0

        attacker_name = attacker.get_name()

        # Make sequence of all potential interceptors from all civs at war
        potential_interceptors = []
        for intercepting_civ in attacker.get_civ_info().game_info.civilizations:
            if attacker.get_civ_info().is_at_war_with(intercepting_civ):
                potential_interceptors.extend(
                    unit for unit in intercepting_civ.units.get_civ_units()
                    if unit.can_intercept(attacked_tile)
                )

        # First priority: only air units
        if any(unit.base_unit.is_air_unit() for unit in potential_interceptors):
            potential_interceptors = [
                unit for unit in potential_interceptors
                if unit.base_unit.is_air_unit()
            ]

        # Pick highest chance interceptor
        for interceptor in sorted(
            random.sample(potential_interceptors, len(potential_interceptors)),
            key=lambda x: x.intercept_chance(),
            reverse=True
        ):
            interceptor.attacks_this_turn += 1

            if not interceptor.base_unit.is_air_unit():
                interceptor_name = interceptor.name
                attacker_text = f"Our [{attacker_name}] ([-0] HP) was attacked by an intercepting [{interceptor_name}] ([-0] HP)"
                interceptor_text = f"Our [{interceptor_name}] ([-0] HP) intercepted and attacked an enemy [{attacker_name}] ([-0] HP)"

                attacker.get_civ_info().add_notification(
                    attacker_text,
                    [(interceptor.current_tile.position, attacker.unit.current_tile.position)],
                    "War",
                    attacker_name,
                    "War",
                    interceptor_name
                )

                interceptor.civ.add_notification(
                    interceptor_text,
                    [(interceptor.current_tile.position, attacker.unit.current_tile.position)],
                    "War",
                    interceptor_name,
                    "War",
                    attacker_name
                )

                attacker.unit.action = None
                return

            # Damage if Air v Air should work similar to Melee
            damage_dealt = Battle.take_damage(attacker, MapUnitCombatant(interceptor))

            # 5 XP to both
            Battle.add_xp(MapUnitCombatant(interceptor), 5, attacker)
            Battle.add_xp(attacker, 5, MapUnitCombatant(interceptor))

            locations_interceptor_unknown = [
                (attacked_tile.position, attacker.unit.current_tile.position)
            ]
            locations = [
                (interceptor.current_tile.position, attacker.unit.current_tile.position)
            ]

            AirInterception._add_air_sweep_interception_notifications(
                attacker,
                interceptor,
                damage_dealt,
                locations_interceptor_unknown,
                locations
            )
            attacker.unit.action = None
            return

        # No interceptions available
        attacker_text = f"Nothing tried to intercept our [{attacker_name}]"
        attacker.get_civ_info().add_notification(
            attacker_text,
            "War",
            attacker_name
        )
        attacker.unit.action = None

    @staticmethod
    def try_intercept_air_attack(
        attacker: MapUnitCombatant,
        attacked_tile,
        intercepting_civ,
        defender: Optional[MapUnitCombatant]
    ) -> DamageDealt:
        if attacker.unit.has_unique(
            "CannotBeIntercepted",
            StateForConditionals(
                attacker.get_civ_info(),
                our_combatant=attacker,
                their_combatant=defender,
                attacked_tile=attacked_tile
            )
        ):
            return DamageDealt.none()

        # Pick highest chance interceptor
        interceptors = [
            unit for unit in intercepting_civ.units.get_civ_units()
            if unit.can_intercept(attacked_tile)
        ]

        interceptors.sort(key=lambda x: x.intercept_chance(), reverse=True)

        for unit in interceptors:
            conditional_state = StateForConditionals(
                intercepting_civ,
                our_combatant=MapUnitCombatant(unit),
                their_combatant=attacker,
                combat_action="Intercept",
                attacked_tile=attacked_tile
            )

            if (unit.get_matching_uniques("CannotInterceptUnits", conditional_state)
                and unit != (defender.unit if isinstance(defender, MapUnitCombatant) else None)):
                continue

            interceptor = unit
            break
        else:
            return DamageDealt.none()

        interceptor.attacks_this_turn += 1

        # Does intercept happen?
        if random.random() > interceptor.intercept_chance() / 100.0:
            return DamageDealt.none()

        damage = BattleDamage.calculate_damage_to_defender(
            MapUnitCombatant(interceptor),
            attacker
        )

        damage_factor = 1.0 + interceptor.intercept_damage_percent_bonus() / 100.0
        damage_factor *= attacker.unit.received_intercept_damage_factor()

        damage = min(
            int(damage * damage_factor),
            attacker.unit.health
        )

        attacker.take_damage(damage)
        if damage > 0:
            Battle.add_xp(MapUnitCombatant(interceptor), 2, attacker)

        AirInterception._add_interception_notifications(attacker, interceptor, damage)

        return DamageDealt(0, damage)

    @staticmethod
    def _add_air_sweep_interception_notifications(
        attacker: MapUnitCombatant,
        interceptor,
        damage_dealt: DamageDealt,
        locations_interceptor_unknown: Sequence[Tuple],
        locations: Sequence[Tuple]
    ) -> None:
        attacker_name = attacker.get_name()
        interceptor_name = interceptor.name

        attacker_text = (
            f"Our [{attacker_name}] ([-{damage_dealt.defender_dealt}] HP) was destroyed by an intercepting [{interceptor_name}] ([-{damage_dealt.attacker_dealt}] HP)"
            if attacker.is_defeated() and interceptor.get_tile() in attacker.get_civ_info().viewable_tiles
            else f"Our [{attacker_name}] ([-{damage_dealt.defender_dealt}] HP) was destroyed by an unknown interceptor"
            if attacker.is_defeated()
            else f"Our [{attacker_name}] ([-{damage_dealt.defender_dealt}] HP) destroyed an intercepting [{interceptor_name}] ([-{damage_dealt.attacker_dealt}] HP)"
            if MapUnitCombatant(interceptor).is_defeated()
            else f"Our [{attacker_name}] ([-{damage_dealt.defender_dealt}] HP) was attacked by an intercepting [{interceptor_name}] ([-{damage_dealt.attacker_dealt}] HP)"
        )

        attacker.get_civ_info().add_notification(
            attacker_text,
            locations_interceptor_unknown,
            "War",
            attacker_name,
            "War",
            "Question"
        )

        interceptor_text = (
            f"Our [{interceptor_name}] ([-{damage_dealt.attacker_dealt}] HP) intercepted and destroyed an enemy [{attacker_name}] ([-{damage_dealt.defender_dealt}] HP)"
            if attacker.is_defeated()
            else f"Our [{interceptor_name}] ([-{damage_dealt.attacker_dealt}] HP) intercepted and was destroyed by an unknown enemy"
            if MapUnitCombatant(interceptor).is_defeated() and attacker.get_tile() not in interceptor.civ.viewable_tiles
            else f"Our [{interceptor_name}] ([-{damage_dealt.attacker_dealt}] HP) intercepted and was destroyed by an enemy [{attacker_name}] ([-{damage_dealt.defender_dealt}] HP)"
            if MapUnitCombatant(interceptor).is_defeated()
            else f"Our [{interceptor_name}] ([-{damage_dealt.attacker_dealt}] HP) intercepted and attacked an enemy [{attacker_name}] ([-{damage_dealt.defender_dealt}] HP)"
        )

        interceptor.civ.add_notification(
            interceptor_text,
            locations,
            "War",
            interceptor_name,
            "War",
            attacker_name
        )

    @staticmethod
    def _add_interception_notifications(
        attacker: MapUnitCombatant,
        interceptor,
        damage: int
    ) -> None:
        attacker_name = attacker.get_name()
        interceptor_name = interceptor.name

        locations = [(interceptor.current_tile.position, attacker.unit.current_tile.position)]

        attacker_text = (
            f"Our [{attacker_name}] ([-{damage}] HP) was attacked by an intercepting [{interceptor_name}] ([-0] HP)"
            if not attacker.is_defeated()
            else f"Our [{attacker_name}] ([-{damage}] HP) was destroyed by an intercepting [{interceptor_name}] ([-0] HP)"
            if interceptor.get_tile() in attacker.get_civ_info().viewable_tiles
            else f"Our [{attacker_name}] ([-{damage}] HP) was destroyed by an unknown interceptor"
        )

        attacker.get_civ_info().add_notification(
            attacker_text,
            interceptor.current_tile.position,
            "War",
            attacker_name,
            "War",
            interceptor_name
        )

        interceptor_text = (
            f"Our [{interceptor_name}] ([-0] HP) intercepted and destroyed an enemy [{attacker_name}] ([-{damage}] HP)"
            if attacker.is_defeated()
            else f"Our [{interceptor_name}] ([-0] HP) intercepted and attacked an enemy [{attacker_name}] ([-{damage}] HP)"
        )

        interceptor.civ.add_notification(
            interceptor_text,
            locations,
            "War",
            interceptor_name,
            "War",
            attacker_name
        )

File: unciv-rust-port/src/battle/air_interception.rs
use rand::Rng;

pub struct AirInterception;

impl AirInterception {
    /// Should draw an Interception if available on the tile from any Civ
    /// Land Units deal 0 damage, and no XP for either party
    /// Air Interceptors do Air Combat as if Melee (mutual damage) but using Ranged Strength. 5XP to both
    /// But does not use the Interception mechanic bonuses/promotions
    /// Counts as an Attack for both units
    /// Will always draw out an Interceptor's attack (they cannot miss)
    /// This means the combat against Air Units will execute and always deal damage
    /// Random Civ at War will Intercept, prioritizing Air Units,
    /// sorted by highest Intercept chance (same as regular Intercept)
    pub fn air_sweep(attacker: &mut MapUnitCombatant, attacked_tile: &Tile) {
        // Air Sweep counts as an attack, even if nothing else happens
        attacker.unit.attacks_this_turn += 1;

        // copied and modified from reduceAttackerMovementPointsAndAttacks()
        // use up movement
        if attacker.unit.has_unique(UniqueType::CanMoveAfterAttacking) ||
           attacker.unit.max_attacks_per_turn() > attacker.unit.attacks_this_turn {
            // if it was a melee attack and we won, then the unit ALREADY got movement points deducted,
            // for the movement to the enemy's tile!
            // and if it's an air unit, it only has 1 movement anyway, so...
            if !attacker.unit.base_unit.moves_like_air_units {
                attacker.unit.use_movement_points(1.0);
            }
        } else {
            attacker.unit.current_movement = 0.0;
        }

        let attacker_name = attacker.get_name();

        // Make giant sequence of all potential Interceptors from all Civs isAtWarWith()
        let mut potential_interceptors = Vec::new();
        for intercepting_civ in attacker.get_civ_info().game_info.civilizations.iter()
            .filter(|c| attacker.get_civ_info().is_at_war_with(c)) {
            potential_interceptors.extend(
                intercepting_civ.units.get_civ_units()
                    .filter(|u| u.can_intercept(attacked_tile))
            );
        }

        // first priority, only Air Units
        if potential_interceptors.iter().any(|u| u.base_unit.is_air_unit()) {
            potential_interceptors.retain(|u| u.base_unit.is_air_unit());
        }

        // Pick highest chance interceptor
        for interceptor in potential_interceptors.iter()
            .collect::<Vec<_>>()
            .shuffle(&mut rand::thread_rng())
            .into_iter()
            .sorted_by(|a, b| b.intercept_chance().partial_cmp(&a.intercept_chance()).unwrap_or(std::cmp::Ordering::Equal)) {

            // No chance of Interceptor to miss (unlike regular Interception). Always want to deal damage
            // pairs of LocationAction for Notification
            let locations = LocationAction::new(
                interceptor.current_tile.position,
                attacker.unit.current_tile.position
            );

            interceptor.attacks_this_turn += 1;  // even if you miss, you took the shot

            if !interceptor.base_unit.is_air_unit() {
                let interceptor_name = interceptor.name.clone();
                // Deal no damage (moddable in future?) and no XP
                let attacker_text = format!(
                    "Our [{}] ([-0] HP) was attacked by an intercepting [{}] ([-0] HP)",
                    attacker_name, interceptor_name
                );
                let interceptor_text = format!(
                    "Our [{}] ([-0] HP) intercepted and attacked an enemy [{}] ([-0] HP)",
                    interceptor_name, attacker_name
                );

                attacker.get_civ_info().add_notification(
                    &attacker_text,
                    &locations,
                    NotificationCategory::War,
                    &attacker_name,
                    NotificationIcon::War,
                    &interceptor_name
                );

                interceptor.civ.add_notification(
                    &interceptor_text,
                    &locations,
                    NotificationCategory::War,
                    &interceptor_name,
                    NotificationIcon::War,
                    &attacker_name
                );

                attacker.unit.action = None;
                return;
            }

            // Damage if Air v Air should work similar to Melee
            let damage_dealt = Battle::take_damage(attacker, MapUnitCombatant::new(interceptor));

            // 5 XP to both
            Battle::add_xp(MapUnitCombatant::new(interceptor), 5, attacker);
            Battle::add_xp(attacker, 5, MapUnitCombatant::new(interceptor));

            let locations_interceptor_unknown = LocationAction::new(
                attacked_tile.position,
                attacker.unit.current_tile.position
            );

            Self::add_air_sweep_interception_notifications(
                attacker,
                interceptor,
                damage_dealt,
                &locations_interceptor_unknown,
                &locations
            );

            attacker.unit.action = None;
            return;
        }

        // No Interceptions available
        let attacker_text = format!("Nothing tried to intercept our [{}]", attacker_name);
        attacker.get_civ_info().add_notification(
            &attacker_text,
            NotificationCategory::War,
            &attacker_name
        );

        attacker.unit.action = None;
    }

    // TODO: Check overlap with addInterceptionNotifications, and unify what we can
    fn add_air_sweep_interception_notifications(
        attacker: &MapUnitCombatant,
        interceptor: &MapUnit,
        damage_dealt: DamageDealt,
        locations_interceptor_unknown: &LocationAction,
        locations: &LocationAction
    ) {
        let attacker_name = attacker.get_name();
        let interceptor_name = interceptor.name.clone();

        let attacker_text = if attacker.is_defeated() {
            if attacker.get_civ_info().viewable_tiles.contains(&interceptor.get_tile()) {
                format!(
                    "Our [{}] ([-{}] HP) was destroyed by an intercepting [{}] ([-{}] HP)",
                    attacker_name, damage_dealt.defender_dealt, interceptor_name, damage_dealt.attacker_dealt
                )
            } else {
                format!(
                    "Our [{}] ([-{}] HP) was destroyed by an unknown interceptor",
                    attacker_name, damage_dealt.defender_dealt
                )
            }
        } else if MapUnitCombatant::new(interceptor).is_defeated() {
            format!(
                "Our [{}] ([-{}] HP) destroyed an intercepting [{}] ([-{}] HP)",
                attacker_name, damage_dealt.defender_dealt, interceptor_name, damage_dealt.attacker_dealt
            )
        } else {
            format!(
                "Our [{}] ([-{}] HP) was attacked by an intercepting [{}] ([-{}] HP)",
                attacker_name, damage_dealt.defender_dealt, interceptor_name, damage_dealt.attacker_dealt
            )
        };

        attacker.get_civ_info().add_notification(
            &attacker_text,
            locations_interceptor_unknown,
            NotificationCategory::War,
            &attacker_name,
            NotificationIcon::War,
            NotificationIcon::Question
        );

        let interceptor_text = if attacker.is_defeated() {
            format!(
                "Our [{}] ([-{}] HP) intercepted and destroyed an enemy [{}] ([-{}] HP)",
                interceptor_name, damage_dealt.attacker_dealt, attacker_name, damage_dealt.defender_dealt
            )
        } else if MapUnitCombatant::new(interceptor).is_defeated() {
            if interceptor.civ.viewable_tiles.contains(&attacker.get_tile()) {
                format!(
                    "Our [{}] ([-{}] HP) intercepted and was destroyed by an enemy [{}] ([-{}] HP)",
                    interceptor_name, damage_dealt.attacker_dealt, attacker_name, damage_dealt.defender_dealt
                )
            } else {
                format!(
                    "Our [{}] ([-{}] HP) intercepted and was destroyed by an unknown enemy",
                    interceptor_name, damage_dealt.attacker_dealt
                )
            }
        } else {
            format!(
                "Our [{}] ([-{}] HP) intercepted and attacked an enemy [{}] ([-{}] HP)",
                interceptor_name, damage_dealt.attacker_dealt, attacker_name, damage_dealt.defender_dealt
            )
        };

        interceptor.civ.add_notification(
            &interceptor_text,
            locations,
            NotificationCategory::War,
            &interceptor_name,
            NotificationIcon::War,
            &attacker_name
        );
    }

    pub fn try_intercept_air_attack(
        attacker: &mut MapUnitCombatant,
        attacked_tile: &Tile,
        intercepting_civ: &Civilization,
        defender: Option<&dyn ICombatant>
    ) -> DamageDealt {
        if attacker.unit.has_unique(
            UniqueType::CannotBeIntercepted,
            StateForConditionals::new(
                attacker.get_civ_info(),
                attacker,
                defender,
                None,
                Some(attacked_tile)
            )
        ) {
            return DamageDealt::None;
        }

        // Pick highest chance interceptor
        let interceptor = intercepting_civ.units.get_civ_units()
            .filter(|u| u.can_intercept(attacked_tile))
            .sorted_by(|a, b| b.intercept_chance().partial_cmp(&a.intercept_chance()).unwrap_or(std::cmp::Ordering::Equal))
            .find(|unit| {
                // Can't intercept if we have a unique preventing it
                let conditional_state = StateForConditionals::new(
                    intercepting_civ,
                    MapUnitCombatant::new(unit),
                    attacker,
                    Some(CombatAction::Intercept),
                    Some(attacked_tile)
                );

                unit.get_matching_uniques(UniqueType::CannotInterceptUnits, conditional_state)
                    .iter()
                    .none(|u| attacker.matches_filter(&u.params[0]))
                    // Defender can't intercept either
                    && unit != defender.and_then(|d| d.downcast_ref::<MapUnitCombatant>()).map(|d| &d.unit).unwrap_or(&None)
            });

        let interceptor = match interceptor {
            Some(i) => i,
            None => return DamageDealt::None,
        };

        interceptor.attacks_this_turn += 1;  // even if you miss, you took the shot

        // Does Intercept happen? If not, exit
        if rand::thread_rng().gen::<f32>() > interceptor.intercept_chance() / 100.0 {
            return DamageDealt::None;
        }

        let mut damage = BattleDamage::calculate_damage_to_defender(
            MapUnitCombatant::new(interceptor),
            attacker
        );

        let mut damage_factor = 1.0 + interceptor.intercept_damage_percent_bonus() as f32 / 100.0;
        damage_factor *= attacker.unit.received_intercept_damage_factor();

        damage = (damage as f32 * damage_factor).min(attacker.unit.health as f32) as i32;

        attacker.take_damage(damage);
        if damage > 0 {
            Battle::add_xp(MapUnitCombatant::new(interceptor), 2, attacker);
        }

        Self::add_interception_notifications(attacker, interceptor, damage);

        DamageDealt::new(0, damage)
    }

    fn add_interception_notifications(
        attacker: &MapUnitCombatant,
        interceptor: &MapUnit,
        damage: i32
    ) {
        let attacker_name = attacker.get_name();
        let interceptor_name = interceptor.name.clone();

        let locations = LocationAction::new(
            interceptor.current_tile.position,
            attacker.unit.current_tile.position
        );

        let attacker_text = if !attacker.is_defeated() {
            format!(
                "Our [{}] ([-{}] HP) was attacked by an intercepting [{}] ([-0] HP)",
                attacker_name, damage, interceptor_name
            )
        } else if attacker.get_civ_info().viewable_tiles.contains(&interceptor.get_tile()) {
            format!(
                "Our [{}] ([-{}] HP) was destroyed by an intercepting [{}] ([-0] HP)",
                attacker_name, damage, interceptor_name
            )
        } else {
            format!(
                "Our [{}] ([-{}] HP) was destroyed by an unknown interceptor",
                attacker_name, damage
            )
        };

        attacker.get_civ_info().add_notification(
            &attacker_text,
            &interceptor.current_tile.position,
            NotificationCategory::War,
            &attacker_name,
            NotificationIcon::War,
            &interceptor_name
        );

        let interceptor_text = if attacker.is_defeated() {
            format!(
                "Our [{}] ([-0] HP) intercepted and destroyed an enemy [{}] ([-{}] HP)",
                interceptor_name, attacker_name, damage
            )
        } else {
            format!(
                "Our [{}] ([-0] HP) intercepted and attacked an enemy [{}] ([-{}] HP)",
                interceptor_name, attacker_name, damage
            )
        };

        interceptor.civ.add_notification(
            &interceptor_text,
            &locations,
            NotificationCategory::War,
            &interceptor_name,
            NotificationIcon::War,
            &attacker_name
        );
    }
}